### PR TITLE
support security scopes in m4

### LINKED ISSUE #1322: Change credential-scopes to security-scopes
`credential-scopes` is defined by ourself. Autorest has provided authentication related definition and it uses `secuirty-scopes`. The details can be found in https://github.com/Azure/autorest/blob/main/docs/generate/authentication.md

### LINKED ISSUE #1356: Support the KeyCredential in generatedClient.ts
In RLC we support three types of authentications:

- Only token credential ==> supported
- Only key credential  ==> issue found
- Both key credential and token credential ==> supported

We have bugs in the second case which if user defined the `credential-key-header-name: x-api-key` the generated factory 
would miss the parameter `credentials`.

```typescript
export default function createClient(
  endpoint: string,
  options: ClientOptions = {}
):
```

Should be like:
```typescript
export default function createClient(
  endpoint: string,
  credentials: KeyCredential,
  options: ClientOptions = {}
):
```

### LINKED ISSUE #1385: Generate authentication support from REST API security definitions
**Describe the solution you'd like**

The TypeScript/JavaScript generator should support generation of authentication support based on the security definitions in the REST API document.

**Additional context**

In the March 2022 release, Autorest generalized its support for recognizing security schemes:

https://github.com/Azure/autorest/blob/main/docs/releases/march-2022.md#modelerfour-security-scheme-generalization

This makes it possible for generators for each language to generate proper authentication support in client libraries for Azure using only information from the standard security definitions in the REST API document.

In particular:
- If the REST API document contains an "oauth2" security scheme, the generated client library should support for authentication with AAD Token credentials.
- If the REST API document contains an "apiKey" security scheme with "in: header", the generated client library should support authentication with an api key.
- If both "oauth2" and "apiKey" security schemes are present, the generated library should support authentication with _either_ AAD Token credentials or an api key.

### PR DESCRIPTION
**Pull request by qiaozha**
fix #1385
fix #1356
fix #1322

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -70,6 +70,7 @@ modelerfour:
   group-parameters: true
   # Enable older inconsistent behavior that an enum with a single value would become a constant by default.
   seal-single-value-enum-by-default: true
+  legacy-request-body: true
```
Comment (qiaozha) [inline README.md:73]: create an issue to track removal of this option #1394

File: test/integration/generated/datafactory/src/dataFactoryClient.ts [hunk 2]
```diff
@@ -82,6 +82,9 @@ export class DataFactoryClient extends coreClient.ServiceClient {
         ? `${options.userAgentOptions.userAgentPrefix} ${packageDetails}`
         : `${packageDetails}`;
 
+    if (!options.credentialScopes) {
+      options.credentialScopes = ["user_impersonation"];
```
Comment (qiaozha) [inline test/integration/generated/datafactory/src/dataFactoryClient.ts:86]: This is coming from the code model, I wonder whether we need to filter this ?

File: test/integration/generated/healthcareapis/src/healthCareApisClient.ts [hunk 3]
```diff
@@ -64,9 +64,6 @@ export class HealthCareApisClient extends coreClient.ServiceClient {
         ? `${options.userAgentOptions.userAgentPrefix} ${packageDetails}`
         : `${packageDetails}`;
 
-    if (!options.credentialScopes) {
-      options.credentialScopes = ["https://management.azure.com/.default"];
-    }
```
Comment (qiaozha) [inline test/integration/generated/healthcareapis/src/healthCareApisClient.ts:69]: There's an add-credential: false in the swagger config, I guess removing it is the correct behavior.

File: src/generators/clientFileGenerator.ts [hunk 4]
```diff
@@ -628,14 +637,19 @@ function writeDefaultOptions(
...
+  const { credentialScopes } = getSecurityInfoFromModel(clientDetails.security);
 
   const credentialScopesValues = getCredentialScopesValue(credentialScopes);
-  const addScopes = credentialScopes
-    ? `if(!options.credentialScopes) {
+  const addScopes =
+    addCredentials &&
+    credentialScopes &&
+    credentialScopesValues &&
+    credentialScopes.length > 0
```
Comment (joheredi) [inline src/generators/clientFileGenerator.ts:648]: I wonder if we can have a function for this, as it is getting bigger

File: src/transforms/objectTransforms.ts [hunk 5]
```diff
@@ -58,7 +58,11 @@ export function transformObject(
     hasAdditionalProperties: false,
     kind,
     name,
-    serializedName: metadata.serializedName,
+    serializedName: metadata.serializedName
+      ? metadata.serializedName
+      : kind === ObjectKind.Polymorphic
+      ? metadata.name
+      : undefined,
```
Comment (joheredi) [inline src/transforms/objectTransforms.ts:65]: Can we also extract these checks into a function to make it easier to read?

File: test/smoke/generated/purview-administration-rest/src/metadataPolicies/purviewMetadataPolicies.ts [hunk 6]
```diff
@@ -2,20 +2,20 @@
...
+  credentials: KeyCredential,
   options: ClientOptions = {}
 ): PurviewMetadataPoliciesClient {
   const baseUrl = options.baseUrl ?? `${Endpoint}/policyStore`;
   options.apiVersion = options.apiVersion ?? "2021-07-01-preview";
   options = {
     ...options,
     credentials: {
-      scopes: ["https://purview.azure.net/.default"]
+      apiKeyHeaderName: "CustomAuth"
```
Comment (qiaozha) [inline test/smoke/generated/purview-administration-rest/src/metadataPolicies/purviewMetadataPolicies.ts:18]: @joheredi  Is this the correct behavior for apiKey credential scenario ?

File: src/restLevelClient/generateClient.ts [hunk 7]
```diff
@@ -42,20 +45,37 @@ export function generateClient(model: CodeModel, project: Project) {
...
+      ? [
+          {
+            name: uriParameter,
+            type: "string"
+          }
+        ]
+      : []),
+    ...(addCredentials === false ||
+    ((!credentialScopes || credentialScopes.length === 0) &&
+      !credentialKeyHeaderName)
```
Comment (joheredi) [inline src/restLevelClient/generateClient.ts:72]: Can we extract this check to a function?

**Review by joheredi — APPROVED** [review]

**Review by sarangan12 — APPROVED** [review]
LGTM

**Pull request merged**